[0.24.0 regression] `@overload`s of `responses.activate` need to be restored too

### Describe the bug

#684 incorrectly removed the `@overload` declarations for `CallList.__getitem__` and `responses.activate`. The former was restored in #690, but the latter needs to be restored too, as shown by the example below.

### Additional context

_No response_

### Version of `responses`

0.24.0

### Steps to Reproduce

Run mypy on this code:

```python
import responses


class C:
    @responses.activate
    def f(self) -> str:
        return "hello"

    def g(self) -> None:
        greeting: str = self.f()
```

### Expected Result

```
Success: no issues found in 1 source file
```

### Actual Result

```
test.py:10: error: Too few arguments  [call-arg]
test.py:10: error: Incompatible types in assignment (expression has type "Callable[..., Any] | Any", variable has type "str")  [assignment]
Found 2 errors in 1 file (checked 1 source file)
```

--- a/CHANGES
+++ b/CHANGES
@@ -1,7 +1,7 @@
 0.24.1
 ------
 
-* Reintroduced overloads for better typing in `CallList`.
+* Reverted overloads removal
 * Added typing to `Call` attributes.
 
 

--- a/mypy.ini
+++ b/mypy.ini
@@ -1,5 +1,4 @@
 [mypy]
-exclude = tests
 show_column_numbers=True
 show_error_codes = True
 
@@ -27,3 +26,8 @@ warn_unreachable=False
 
 strict_equality=True
 ignore_missing_imports=True
+
+[mypy-responses.tests.*]
+disallow_untyped_calls=False
+disallow_untyped_defs=False
+disable_error_code = union-attr

--- a/responses/__init__.py
+++ b/responses/__init__.py
@@ -405,6 +405,9 @@ class BaseResponse:
         self._calls: CallList = CallList()
         self.passthrough = passthrough
 
+        self.status: int = 200
+        self.body: "_Body" = ""
+
     def __eq__(self, other: Any) -> bool:
         if not isinstance(other, BaseResponse):
             return False
@@ -569,6 +572,8 @@ class Response(BaseResponse):
         auto_calculate_content_length: bool = False,
         **kwargs: Any,
     ) -> None:
+        super().__init__(method, url, **kwargs)
+
         # if we were passed a `json` argument,
         # override the body and content_type
         if json is not None:
@@ -596,7 +601,6 @@ class Response(BaseResponse):
         self.stream: Optional[bool] = stream
         self.content_type: str = content_type  # type: ignore[assignment]
         self.auto_calculate_content_length: bool = auto_calculate_content_length
-        super().__init__(method, url, **kwargs)
 
     def get_response(self, request: "PreparedRequest") -> HTTPResponse:
         if self.body and isinstance(self.body, Exception):
@@ -641,6 +645,8 @@ class CallbackResponse(BaseResponse):
         content_type: Optional[str] = "text/plain",
         **kwargs: Any,
     ) -> None:
+        super().__init__(method, url, **kwargs)
+
         self.callback = callback
 
         if stream is not None:
@@ -650,7 +656,6 @@ class CallbackResponse(BaseResponse):
             )
         self.stream: Optional[bool] = stream
         self.content_type: Optional[str] = content_type
-        super().__init__(method, url, **kwargs)
 
     def get_response(self, request: "PreparedRequest") -> HTTPResponse:
         headers = self.get_headers()
@@ -969,6 +974,22 @@ class RequestsMock:
         finally:
             self.reset()
         return success
+
+    @overload
+    def activate(self, func: "_F" = ...) -> "_F":
+        """Overload for scenario when 'responses.activate' is used."""
+
+    @overload
+    def activate(  # type: ignore[misc]
+        self,
+        *,
+        registry: Type[Any] = ...,
+        assert_all_requests_are_fired: bool = ...,
+    ) -> Callable[["_F"], "_F"]:
+        """Overload for scenario when
+        'responses.activate(registry=, assert_all_requests_are_fired=True)' is used.
+        See https://github.com/getsentry/responses/pull/469 for more details
+        """
 
     def activate(
         self,

--- a/responses/_recorder.py
+++ b/responses/_recorder.py
@@ -50,8 +50,8 @@ def _dump(
                     "response": {
                         "method": rsp.method,
                         "url": rsp.url,
-                        "body": rsp.body,  # type: ignore[attr-defined]
-                        "status": rsp.status,  # type: ignore[attr-defined]
+                        "body": rsp.body,
+                        "status": rsp.status,
                         "headers": rsp.headers,
                         "content_type": rsp.content_type,
                         "auto_calculate_content_length": content_length,

--- a/responses/matchers.py
+++ b/responses/matchers.py
@@ -7,6 +7,7 @@ from typing import Callable
 from typing import Dict
 from typing import List
 from typing import Optional
+from typing import Pattern
 from typing import Tuple
 from typing import Union
 from urllib.parse import parse_qsl
@@ -391,7 +392,7 @@ def multipart_matcher(
 
 
 def header_matcher(
-    headers: Dict[str, str], strict_match: bool = False
+    headers: Dict[str, Union[str, Pattern[str]]], strict_match: bool = False
 ) -> Callable[..., Any]:
     """
     Matcher to match 'headers' argument in request using the responses library.

--- a/responses/tests/test_matchers.py
+++ b/responses/tests/test_matchers.py
@@ -1,5 +1,7 @@
 import gzip
 import re
+from typing import Any
+from typing import List
 from unittest.mock import Mock
 
 import pytest
@@ -152,7 +154,7 @@ def test_json_params_matcher_json_list():
 
 
 def test_json_params_matcher_json_list_empty():
-    json_a = []
+    json_a: "List[Any]" = []
     json_b = "[]"
     mock_request = Mock(body=json_b)
     result = matchers.json_params_matcher(json_a)(mock_request)
@@ -457,7 +459,7 @@ def test_fail_matchers_error():
         (b"\xacHello World!", b"\xacHello World!"),
     ],
 )
-def test_multipart_matcher(req_file, match_file):
+def test_multipart_matcher(req_file, match_file):  # type: ignore[misc]
     @responses.activate
     def run():
         req_data = {"some": "other", "data": "fields"}
@@ -796,7 +798,7 @@ def test_matchers_create_key_val_str():
 
 class TestHeaderWithRegex:
     @property
-    def url(self):
+    def url(self):  # type: ignore[misc]
         return "http://example.com/"
 
     def _register(self):

--- a/responses/tests/test_multithreading.py
+++ b/responses/tests/test_multithreading.py
@@ -10,7 +10,7 @@ import responses
 
 
 @pytest.mark.parametrize("execution_number", range(10))
-def test_multithreading_lock(execution_number):
+def test_multithreading_lock(execution_number):  # type: ignore[misc]
     """Reruns test multiple times since error is random and
     depends on CPU and can lead to false positive result.
 

--- a/responses/tests/test_recorder.py
+++ b/responses/tests/test_recorder.py
@@ -97,9 +97,9 @@ class TestRecord:
 
         def dump_to_file(file_path, registered):
             with open(file_path, "wb") as file:
-                _dump(registered, file, tomli_w.dump)
+                _dump(registered, file, tomli_w.dump)  # type: ignore[arg-type]
 
-        custom_recorder.dump_to_file = dump_to_file
+        custom_recorder.dump_to_file = dump_to_file  # type: ignore[method-assign]
 
         url202, url400, url404, url500 = self.prepare_server(httpserver)
 
@@ -151,12 +151,12 @@ class TestReplay:
         assert not self.out_file.exists()
 
     @pytest.mark.parametrize("parser", (yaml, tomli_w))
-    def test_add_from_file(self, parser):
+    def test_add_from_file(self, parser):  # type: ignore[misc]
         if parser == yaml:
             with open(self.out_file, "w") as file:
                 parser.dump(get_data("example.com", "8080"), file)
         else:
-            with open(self.out_file, "wb") as file:
+            with open(self.out_file, "wb") as file:  # type: ignore[assignment]
                 parser.dump(get_data("example.com", "8080"), file)
 
         @responses.activate
@@ -164,12 +164,12 @@ class TestReplay:
             responses.patch("http://httpbin.org")
             if parser == tomli_w:
 
-                def _parse_response_file(file_path):
+                def _parse_resp_f(file_path):
                     with open(file_path, "rb") as file:
                         data = _toml.load(file)
                     return data
 
-                responses.mock._parse_response_file = _parse_response_file
+                responses.mock._parse_response_file = _parse_resp_f  # type: ignore[method-assign]
 
             responses._add_from_file(file_path=self.out_file)
             responses.post("http://httpbin.org/form")

--- a/responses/tests/test_responses.py
+++ b/responses/tests/test_responses.py
@@ -123,7 +123,7 @@ def test_response_with_instance():
         ),
     ],
 )
-def test_replace(original, replacement):
+def test_replace(original, replacement):  # type: ignore[misc]
     @responses.activate
     def run():
         responses.add(responses.GET, "http://example.com/one", body="test1")
@@ -157,7 +157,7 @@ def test_replace(original, replacement):
         (re.compile(r"http://example\.com/one"), "http://example.com/one"),
     ],
 )
-def test_replace_error(original, replacement):
+def test_replace_error(original, replacement):  # type: ignore[misc]
     @responses.activate
     def run():
         responses.add(responses.GET, original)
@@ -203,7 +203,7 @@ def test_replace_response_object_error():
         ),
     ],
 )
-def test_upsert_replace(original, replacement):
+def test_upsert_replace(original, replacement):  # type: ignore[misc]
     @responses.activate
     def run():
         responses.add(responses.GET, "http://example.com/one", body="test1")
@@ -241,7 +241,7 @@ def test_upsert_replace(original, replacement):
         ),
     ],
 )
-def test_upsert_add(original, replacement):
+def test_upsert_add(original, replacement):  # type: ignore[misc]
     @responses.activate
     def run():
         responses.add(responses.GET, "http://example.com/one", body="test1")
@@ -299,7 +299,7 @@ def test_remove():
         ),
     ],
 )
-def test_response_equality(args1, kwargs1, args2, kwargs2, expected):
+def test_response_equality(args1, kwargs1, args2, kwargs2, expected):  # type: ignore[misc]
     o1 = BaseResponse(*args1, **kwargs1)
     o2 = BaseResponse(*args2, **kwargs2)
     assert (o1 == o2) is expected
@@ -847,7 +847,7 @@ class TestAdapters:
         "adapter_class",
         (CustomAdapter, PositionalArgsAdapter, PositionalArgsIncompleteAdapter),
     )
-    def test_custom_adapter(self, adapter_class):
+    def test_custom_adapter(self, adapter_class):  # type: ignore[misc]
         """Test basic adapter implementation and that responses can patch them properly."""
 
         @responses.activate
@@ -902,12 +902,12 @@ def test_activate_doesnt_change_signature():
 
 
 @pytest.fixture
-def my_fruit():
+def my_fruit():  # type: ignore[misc]
     return "apple"
 
 
 @pytest.fixture
-def fruit_basket(my_fruit):
+def fruit_basket(my_fruit):  # type: ignore[misc]
     return ["banana", my_fruit]
 
 
@@ -926,7 +926,7 @@ class TestFixtures:
 
 def test_activate_mock_interaction():
     @patch("sys.stdout")
-    def test_function(mock_stdout):
+    def test_function(mock_stdout):  # type: ignore[misc]
         return mock_stdout
 
     decorated_test_function = responses.activate(test_function)
@@ -1045,7 +1045,7 @@ def test_response_cookies_multiple():
 
 @pytest.mark.parametrize("request_stream", (True, False, None))
 @pytest.mark.parametrize("responses_stream", (True, False, None))
-def test_response_cookies_session(request_stream, responses_stream):
+def test_response_cookies_session(request_stream, responses_stream):  # type: ignore[misc]
     @responses.activate
     def run():
         url = "https://example.com/path"
@@ -1384,13 +1384,13 @@ def test_content_length_error(monkeypatch):
     # Type errors here and on 1250 are ignored because the stubs for requests
     # are off https://github.com/python/typeshed/blob/f8501d33c737482a829c6db557a0be26895c5941
     #   /stubs/requests/requests/packages/__init__.pyi#L1
-    original_init = getattr(urllib3.HTTPResponse, "__init__")  # type: ignore
+    original_init = getattr(urllib3.HTTPResponse, "__init__")
 
     def patched_init(self, *args, **kwargs):
         kwargs["enforce_content_length"] = True
         original_init(self, *args, **kwargs)
 
-    monkeypatch.setattr(urllib3.HTTPResponse, "__init__", patched_init)  # type: ignore
+    monkeypatch.setattr(urllib3.HTTPResponse, "__init__", patched_init)
 
     run()
     assert_reset()
@@ -1919,7 +1919,7 @@ def test_custom_target(monkeypatch):
         "http://example.com/other/path/",
     ),
 )
-def test_request_param(url):
+def test_request_param(url):  # type: ignore[misc]
     @responses.activate
     def run():
         params = {"hello": "world", "example": "params"}
@@ -1962,7 +1962,7 @@ def test_request_param_with_multiple_values_for_the_same_key():
 @pytest.mark.parametrize(
     "url", ("http://example.com", "http://example.com?hello=world")
 )
-def test_assert_call_count(url):
+def test_assert_call_count(url):  # type: ignore[misc]
     @responses.activate
     def run():
         responses.add(responses.GET, url)
@@ -2160,7 +2160,7 @@ def test_fail_request_error():
         ),
     ],
 )
-def test_response_representations(response_params, expected_representation):
+def test_response_representations(response_params, expected_representation):  # type: ignore[misc]
     response = Response(**response_params)
 
     assert str(response) == expected_representation
@@ -2206,7 +2206,7 @@ def test_mocked_responses_list_registered():
         ("http://fizzbuzz/foo", "http://fizzbuzz/foo"),
     ],
 )
-def test_rfc_compliance(url, other_url):
+def test_rfc_compliance(url, other_url):  # type: ignore[misc]
     @responses.activate
     def run():
         responses.add(method=responses.GET, url=url)
@@ -2314,7 +2314,7 @@ class TestStrictWrapper:
         run_not_strict()
 
     @pytest.mark.parametrize("assert_fired", (True, False, None))
-    def test_nested_decorators(self, assert_fired):
+    def test_nested_decorators(self, assert_fired):  # type: ignore[misc]
         """Validate if assert_all_requests_are_fired is applied from the correct function.
 
         assert_all_requests_are_fired must be applied from the function
@@ -2607,7 +2607,7 @@ class TestMaxRetry:
         assert_reset()
 
     @pytest.mark.parametrize("raise_on_status", (True, False))
-    def test_max_retries_exceed(self, raise_on_status):
+    def test_max_retries_exceed(self, raise_on_status):  # type: ignore[misc]
         @responses.activate(registry=registries.OrderedRegistry)
         def run():
             url = "https://example.com"

